Implement a pass for expression-flattening and annotation of Sharing

maclement opened the pull request:
The pass for the annotation counts all variables on right hand sides of function declarations and introduces `let` expressions that makes the sharing of variables explicit.

For example the function
```
twice :: Integer -> Integer
twice x = x + x
```
is transformed into

```
twice :: Integer -> Integer
twice x = let x@0 = x in x@0 + x@0
```
where `x@0` is a fresh variable that is also inserted into the environment. 

The Pass for expression-flattening transforms each function into a "flat" form.
An expression is "flattened" if each function `f` is only applied onto variables or constants.

The following function
```
fun f g x = f (g x)
```
is translated into
```
fun f g x = let z = g x in f z
```
where `z` is a fresh variable.

closes #76

maclement committed:
Implement Pass for sharing analysis #76

maclement committed:
Add documentation #76

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+--
+--   == Translation
+--
+--   All shared variables on right hand sides of function declarations are made
```
```suggestion
--   All shared variables on right-hand sides of function declarations are made
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+-- | This module contains a compiler pass that analyses the right hand sides of
+--   function declaration and introduces @let@ expression with new variables
+--   for each variable that occurs more than once on the right hand sides.
```
```suggestion
-- | This module contains a compiler pass that analyses the right-hand sides of
--   function declaration and introduces @let@ expression with new variables
--   for each variable that occurs more than once on the right-hand sides.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+import           FreeC.Pass
+
+-- | Checks all function declarations if they contain variables that occur
+--   multiple times on the same right hand side.
```
```suggestion
-- | Checks all function declarations if they contain variables that occur
--   multiple times on the same right-hand side.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+                funcDecls')
+
+-- | Checks a function declaration if it contains a variable that occurs
+--   multiple times on the right hand side.
```
```suggestion
-- | Checks a function declaration if it contains a variable that occurs
--   multiple times on the right-hand side.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+        _ <- renameAndDefineVar srcSpan False varPatIdent varPatType
+        return ( bind : binds, subst : substs )
+
+-- | Counts all variable names on right hand sides of expression.
```
```suggestion
-- | Counts all variable names on right-hand sides of expression.
--
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+--   function declaration and introduces @let@ expression with new variables
+--   for each variable that occurs more than once on the right hand sides.
+--
+--   = Examples
+--
+--   == Example 1
+--
+--   > twice :: Integer -> Integer
+--   > twice ( x :: Integer ) = x + x
+--
+--   Should be transformed into
+--
+--   > twice :: Integer -> Maybe Integer
+--   > twice ( x :: Integer ) = let ( y :: Integer ) = x in y + y
+--
+--   == Example 2
+--
+--   data Maybe a = Just a | Nothing
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe ( mx :: Maybe Integer ) = case ( mx :: Maybe Integer ) of
+--   >   { Nothing -> Nothing
+--   >   ; Just x -> Just (x + x)
+--   >   }
+--
+--   Should be transformed into
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe ( mx :: Maybe Integer ) = case ( mx :: Maybe Integer ) of
+--   >   { Nothing -> Nothing
+--   >   ; Just x -> let y = x in Just ( y + y )
+--   >   }
```
The following formatting is more consistent with the documentation of the other passes. I don't think that the definition of the `Maybe` data type needs to be recapitulated.

```suggestion
--   = Examples
--
--   == Example 1
--
--   > twice :: Integer -> Integer
--   > twice (x :: Integer) = x + x
--
--   Should be transformed into
--
--   > twice :: Integer -> Maybe Integer
--   > twice (x :: Integer) = let (y :: Integer) = x in y + y
--
--   == Example 2
--
--   > twiceMaybe :: Maybe Integer -> Maybe Integer
--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
--   >     Nothing -> Nothing;
--   >     Just x  -> Just (x + x)
--   >   }
--
--   Should be transformed into
--
--   > twiceMaybe :: Maybe Integer -> Maybe Integer
--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
--   >     Nothing -> Nothing;
--   >     Just x  -> let y = x in Just (y + y)
--   >   }
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+
+-- | Checks all function declarations if they contain variables that occur
+--   multiple times on the same right hand side.
+--   If that is the case a @let@-expression is introduced that binds the
```
```suggestion
--   If that is the case, a @let@-expression is introduced that binds the
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+
+-- | Checks a function declaration if it contains a variable that occurs
+--   multiple times on the right hand side.
+--   If that is the case a @let@-expression is introduced.
```
```suggestion
--   If that is the case, a @let@-expression is introduced.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+    return (IR.FuncDecl srcSpan ident typeArgs args returnType rhs')
+
+-- | Builds a @let@-expression from the given expression and variable names.
+--   Computes let-bindings from the given variables, composes the resulting
```
```suggestion
--
--   Computes @let@-bindings from the given variables, composes the resulting
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,143 @@
+    return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+            (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into let-bindings where
+--   the variable pattern is a fresh variable and the right hand side is
+--   a variable that occured multiple times on right hand sides.
```
```suggestion
-- | Converts the list containing variables into @let@-bindings where
--   the variable pattern is a fresh variable and the right-hand side is
--   a variable that occurred multiple times on right hand sides.
--
```

just95 committed:
Fix typos #169

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+--   >     Just x  -> let y = x in Just (y + y)
+--   >   }
+--
+--
```
Maybe add something like

```suggestion
--
-- where @y@ is a fresh variable.
--
```

here and in the example above.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+--   == Translation
+--
+--   All shared variables on right-hand sides of function declarations are made
+--   explicit by introducing @let@-expressions.
```
This sounds more like the postcondition to me.
The *Translation* section should provide details for how you decide whether a variable is considered shared or not and where exactly `let`-expressions are introduced.
For instance, if Example 2 is correct, don't forget to mention that `let`-expressions are added for shared variables from `case`-expression alternatives as well.
I was also wondering whether the arguments of lambda expressions don't have to be shared explicitly as well.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+-- | This module contains a compiler pass that analyses the right-hand sides of
+--   function declaration and introduces @let@ expression with new variables
```
```suggestion
--   function declaration and introduces @let@-expression with new variables
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+-- | Checks all function declarations if they contain variables that occur
+--   multiple times on the same right-hand side.
+--   If that is the case, a @let@-expression is introduced that binds the
+--   variables to a fresh ones and replaces the occurrences with the newly
```
```suggestion
--   variables to fresh ones and replaces the occurrences with the newly
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+    (IR.Module srcSpan name imports typeDecls typeSigs pragmas funcDecls) = do
+        funcDecls' <- mapM analyseSharingDecl (funcDecls)
+        return (IR.Module srcSpan name imports typeDecls typeSigs pragmas
+                funcDecls')
```
This can be shortened using record syntax.

```suggestion
sharingAnaylsisPass ast = do
  funcDecls' <- mapM analyseSharingDecl (IR.modFuncDecls ast)
  return ast { IR.modFuncDecls = funcDecls' }
```

just95 commented on src/lib/FreeC/Pipeline.hs:
```diff
@@ -38,6 +39,7 @@ pipeline =
     [typeInferencePass, defineFuncDeclsPass, partialityAnalysisPass]
   , completePatternPass
   , etaConversionPass
+  , sharingAnaylsisPass
```
Why don't we move this pass above the `typeInferencePass` such that we don't have to manage type annotations manually?

maclement replied:
That is a good idea. I wasn't sure where to put it.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet
+    :: IR.Expr -> [ ( IR.VarName, Maybe IR.TypeScheme ) ] -> Converter IR.Expr
```
The formatting suggests that your branch may be based on an an early version of the `floskell` branch. Make sure to merge `main` into your branch to get the latest version of `floskell.json` and reformat the code such that the pipeline does not fail when this branch is merged.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+countVarNames (IR.TypeAppExpr _ lhs _ _) = countVarNames lhs
+countVarNames (IR.If _ e1 e2 e3 _) = (countVarNames e1)
+    `mergeMap` (countVarNames e2) `mergeMap` countVarNames e3
+countVarNames (IR.Case _ e alts _) = countVarNames e `mergeMap` foldr mergeMap
+    Map.empty (map (countVarNames . IR.altRhs) alts)
+countVarNames IR.Undefined {} = Map.empty
+countVarNames IR.ErrorExpr {} = Map.empty
+countVarNames IR.IntLiteral {} = Map.empty
+countVarNames (IR.Lambda _ _ rhs _) = countVarNames rhs
+countVarNames (IR.Let _ binds e _) = countVarNames e `mergeMap` foldr mergeMap
+    Map.empty (map (countVarNames . IR.bindExpr) binds)
```
`case`-expressions alternatives, lambda expressions and `let`-bindings can shadow variables. Shadowed variables should probably not be counted.

maclement replied:
Good catch. I didn't think of that.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+countVarNames (IR.App _ lhs rhs _)
+    = countVarNames lhs `mergeMap` countVarNames rhs
+countVarNames (IR.TypeAppExpr _ lhs _ _) = countVarNames lhs
+countVarNames (IR.If _ e1 e2 e3 _) = (countVarNames e1)
+    `mergeMap` (countVarNames e2) `mergeMap` countVarNames e3
```
Looks like the CI pipeline has not been triggered for some reason. HLint should have complained about the redundant parentheses.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+--
+--   Should be transformed into
+--
+--   > twiceMaybe :: Maybe Integer -> Maybe Integer
+--   > twiceMaybe (mx :: Maybe Integer) = case (mx :: Maybe Integer) of {
+--   >     Nothing -> Nothing;
+--   >     Just x  -> let y = x in Just (y + y)
+--   >   }
```
If I understand the code correctly, this is not implemented yet.
However, I'm also not convinced that this is semantically correct. Consider the following program.

```haskell
let m42 = Just (trace "Test" 42) in (twiceMaybe m42, twiceMaybe m42)
```

This program should print "Test" only once but I think that it will be printed twice in the example because `y` is shared only locally.

macmaccamcam replied:
You are right. I think misunderstood the task here.

maclement replied:
You are right. I overlooked to implement introduction of `let`-expressions on right hand sides of `case`-alternatives.

nbun replied:
@just95 If I understand your example correctly, the `let m42` binding introduces sharing, which is not "overwritten" by the binding in `twiceMaybe`. So it does not make a difference whether there is a `let` clause in the right-hand side of the `Just` alternative.

just95 replied:
@nbun Ah, okay, I somehow thought `share` would have no effect on the computation inside `Just`. I've now played around with the Coq implementation and can confirm that "Test" is printed only once in my example.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+    buildBind ( binds, substs ) ( varName, varTypeScheme ) = do
+        varName' <- freshHaskellQName varName
+        let subst = singleSubst' varName (\s -> IR.Var s varName' varTypeScheme)
+            -- This type scheme has wrong srcSpanInformation
```
```suggestion
            -- This type scheme has wrong source span information.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [ ( IR.VarName, Maybe IR.TypeScheme ) ]
+    -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+buildBinds srcSpan = foldM buildBind ( [], [] )
+  where
+    buildBind :: ( [ IR.Bind ], [ Subst IR.Expr ] )
+        -> ( IR.VarName, Maybe IR.TypeScheme )
+        -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
```
I think this fold is just [`mapAndUnzipM`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Control-Monad.html#v:mapAndUnzipM) from `Control.Monad`.

```suggestion
buildBinds srcSpan = mapAndUnzipM buildBind
  where
    buildBind
        :: (IR.VarName, Maybe IR.TypeScheme)
        -> Converter (IR.Bind, Subst IR.Expr)
```

just95 committed:
Fix typos #169

Daniel-Teut commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,144 @@
+            varPatType = fmap IR.typeSchemeType varTypeScheme
+            varPat = IR.VarPat srcSpan varPatIdent varPatType False
+            bind = IR.Bind srcSpan varPat rhs
+        _ <- renameAndDefineVar srcSpan False varPatIdent varPatType
```
Currently only lets defined in this pass are added to the environment and lets already present in the source code are not.  In #179 all variables in let bindings are added during the translation to LIR. I think it won't be necessary to add them here.

maclement replied:
Okay. I was unsure if the variables were added to environment somewhere else.

Daniel-Teut reviewed (changes requested):
If the haskell program already contains let expressions for a shared variable, e.g.
```haskell
f x = let x' = x in x' + x' 
```
a new let is still introduced:
```
f x = let { x'@0 = x' } in (let { x'  = x } in ( x' +  x' )) 
```

maclement committed:
Implement review suggestions #76

maclement committed:
Merge main into issue-76 #76

maclement committed:
Apply floskell #76

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,254 @@
+--
+--   After all subexpressions are checked the right hand side of the function
+--   declaration is checked as well.
+--   Variables already bound by @let@-bindings are not counted.
```
```suggestion
--   == Translation
--
--   First all subexpressions of each function declaration are checked for variables
--   occurring multiple times on the right hand sides of lambda abstractions or
--   @case@-expression alternatives. If found the variables are replaced by a
--   fresh variable and a @let@-binding is introduced binding the fresh variable
--   to the old one.
--
--   After all subexpressions are checked the right hand side of the function
--   declaration is checked as well.
--   Variables already bound by @let@-bindings are not counted.
--
--   == Postconditions
--
--   All shared variables on right-hand sides of function declarations are made
--   explicit by introducing @let@-expressions.
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,254 @@
+    varName' <- freshHaskellQName varName
+    let subst            = singleSubst' varName
+          (\s -> IR.Var s varName' Nothing)
+        -- This type scheme has wrong source span information.
```
There is no type annotation anymore.

```suggestion
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,254 @@
+--   substitutions and applies the substitution on the expression.
+buildLet :: IR.Expr -> [IR.VarName] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
```
Due to the usage of `freshHaskellQName` in `buildBinds` the counter for fresh variables is incremented. Since the scope of the fresh variables is delimited by the `let` expression, I think that we can open a local environment such that the counter is reset afterwards.

```suggestion
buildLet expr vars = localEnv $ do
```

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,254 @@
+countVarNamesExcept :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesExcept = countVarNamesWith (++) elem
+
+-- | Counts all variable names on right-hand sides of expression that are also
+--   contained by the given list.
+--   Shadowed variables and variables from the list are not counted.
+--   Variables introduced on the left side of a @case@-alternative and @let@
+--   expressions are not counted.
+countVarNamesOnly :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesOnly = countVarNamesWith const notElem
```
The documentation states that

> Shadowed variables and variables from the list are not counted

but actually only variables from the list are counted. Shadowed variables are counted 
 as well since `const` ignores bound variables.

just95 commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,254 @@
+--   of @case@-alternatives or lambda abstractions.
+--   The second argument is the a predicate that determines if a variable
+--   is added to the map.
+countVarNamesWith :: ([IR.VarName] -> [IR.VarName] -> [IR.VarName])
+                  -> (IR.VarName -> [IR.VarName] -> Bool)
+                  -> [IR.VarName]
+                  -> IR.Expr
+                  -> Map IR.VarName Integer
```
You've overengineered this function a little bit which makes it very hard to understand. The definition gets much easier without predicates and accumulator. In order to implement shadowing you just have to remove entries from the map, for example.

```haskell
countVarNames :: IR.Expr -> Map IR.VarName Integer
countVarNames (IR.Lambda _ args rhs _) 
  = countVarNames rhs `Map.withoutKeys` Set.fromList (map IR.varPatQName args)
```

The intended behavior of `countVarNamesOnly` can then be obtained by restricting the key set of the map.

```haskell
countVarNamesOnly :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
countVarNamesOnly varNames expr
  = countVarNames expr `Map.restirctKeys` Set.fromList varNames 
```

This fixes the bug pointed out above automatically.

maclement replied:
That is a really good idea. I did not think of that. 
My intention was to avoid code duplication but that made the function way to complicated.

just95 committed:
Apply suggestions from own code review #169

just95 committed:
Fix typo

just95 committed:
Fix bound variables of `let`-expression #76

Daniel-Teut committed:
Fix typos in comments

Daniel-Teut commented on src/lib/FreeC/Pass/SharingAnalysisPass.hs:
```diff
@@ -0,0 +1,253 @@
+-- | Checks a function declaration for @case@-expressions to introduce local
+--   @let@-expressions and applies the transformation on the right-hand side.
+analyseSharingDecl :: IR.FuncDecl -> Converter IR.FuncDecl
+analyseSharingDecl (IR.FuncDecl srcSpan ident typeArgs args returnType rhs) = do
+  rhs' <- analyseLocalSharing rhs
+  rhs'' <- analyseSharingExpr rhs'
+  return (IR.FuncDecl srcSpan ident typeArgs args returnType rhs'')
```
```suggestion
analyseSharingDecl funcDecl = do
  rhs' <- (analyseLocalSharing >=> analyseSharingExpr) (IR.funcDeclRhs funcDecl)
  return funcDecl { IR.funcDeclRhs = rhs' }
```
The function can be simplified a bit.

maclement committed:
Update src/lib/FreeC/Pass/SharingAnalysisPass.hs

Co-authored-by: Justin Andresen <jan.dresen.95@gmail.com>

maclement committed:
Update src/lib/FreeC/Pass/SharingAnalysisPass.hs

Co-authored-by: Daniel Teut <danielteut@t-online.de>

Daniel-Teut committed:
Fix import

just95 committed:
Merge branch 'main' into issue-76

just95 committed:
Don't qualify references to generated `let`-bindings #76

When a `let`-binding is generated during sharing analysis,
the variables for the `let`-binding in the `in`-expression
must not be qualified even if the right-hand side of the
`let`-binding is qualified variable since local variables
are never qualified.

maclement committed:
Apply review suggestions #76

Daniel-Teut committed:
Merge branch 'main' into issue-76

maclement committed:
Fix counting for case alternatives #76

maclement committed:
Merge branch 'issue-76' of https://github.com/FreeProving/free-compiler into issue-76

maclement committed:
Merge branch 'main' into issue-76

maclement committed:
Strip expression types from IR backend #76

maclement committed:
Implement term flattening #76

maclement committed:
re-introduce SharingAnalysisPass #76

maclement committed:
Apply local sharing recursively #76

maclement committed:
Fix typo #76

maclement committed:
Fix tests. #76

Commit:
Add compiler flag for effect encapsulation #211

Commit:
Encapsulate sharing when necessary #211

Commit:
Use varPatQName and format code #211

Commit:
WIP

Commit:
Remove local env #211

Commit:
Encapsulate effects when flattening expressions #211

Commit:
Merge issue-76 into issue-211 #211

Commit:
Apply HLint hint and format code #211

Commit:
Merge with main #211

Commit:
Add Normalform and ShareableArgs instances for Property #211

Commit:
Fix error in getFuncName #211

Commit:
Format code #211

Commit:
Adjust SharingAnalysisPassTests #211

Commit:
Merge branch 'issue-209' of https://github.com/FreeProving/free-compiler into issue-211

Commit:
Fix bug in analysis of case expression variable counting #211

MajaRet committed:
Merge pull request #216 from FreeProving/issue-211

Encapsulate effects in QuickCheck function arguments

maclement committed:
Add documentation to FlattenExprPass #76

maclement committed:
Merge conflicts #76

maclement committed:
Add tracing and more description #76

Commit:
Merge main into issue-76 #76

Commit:
Merge branch 'issue-76' of https://github.com/FreeProving/free-compiler into issue-76

Commit:
Remove Normalform instance for Property #76

maclement committed:
WIP

Daniel-Teut committed:
Fix typos #76

maclement committed:
WIP #76

maclement committed:
Merge locally #76

maclement committed:
WIP #76

Commit:
Add Comb to Free exports #76

Commit:
Adjust example proofs #76

Commit:
Merge branch 'issue-204' of https://github.com/FreeProving/free-compiler into issue-76

Commit:
Adjust proofs #76

Commit:
Resolve merge conflict #76

just95 committed:
Merge branch 'main' into issue-76

just95 committed:
Fix compiler errors from merge with 'main' branch #169

just95 committed:
Make `encapsulates-effects` flag optional #169

just95 committed:
Simplify `FlattenExprPass` #169

just95 committed:
Rename `Tracing.v` to `TracingProofs.v`

just95 committed:
Use call-by-name in razor case study #169

The razor case study was adapted such that the call-by-name
evaluation strategy is used in the proofs.

just95 committed:
Disable `SharingAnalysisPass` #169

When the sharing analysis pass is enabled, no `Section`s are
created by the Coq back end because all arguments are bound by
`let`-expressions.

just95 committed:
Disable decreasing argument example #169

The `mapRose` example does not work anymore in Coq because
Coq does not know how `call` is implemented. Only when the
concrete evaluation strategy is specified Coq can verify
that the function terminates.

just95 committed:
Remove `Injectable` constraint #169

just95 committed:
Annotate decreasing argument in generated type class instances #169

just95 committed:
Inline lambda abstractions bound by `let`-expressions #169

Coq currently does not accept the `mapRose` example because it
does not know how the lambda abstraction (that is introduced by
η-conversion) is invoked. Coq can recognize that the function
terminates if we remove the corresponding `call` operations (i.e.,
`let`-bindings in IR). This does not change the semantics of the
program since lambda abstractions are not evaluated until they are
invoked in every evaluation strategy.

just95 committed:
Use data type instead of type class for `Strategy` #169

just95 committed:
Format code

just95 committed:
Use environment files in CI pipeline

just95 committed:
Fix permissions for writing to environment file

just95 committed:
Update GitHub Actions dependencies

just95 committed:
Update to Cabal 3.4

just95 committed:
Fix decreasing argument analysis for transitive aliases

just95 committed:
Add field for strict arguments to module interface files

just95 committed:
Remove `call` operator #169

The `call` operator was an optimization to avoid the generation of
unnecessary `share`s which would complicate proofs in a call-by-need
setting. However, even if we transform expressions into a flat form
additional `share` operators are needed to model sharing correctly
if we are using the `call` operator. Since we only want to generate
the flat form, we cannot apply the optimization or have to find a
way to model sharing that also works in a call-by-value setting.

just95 committed:
Merge branch 'inductive-strategy' into no-call

just95 committed:
Remove `SharingAnalysisPass` #169

just95 committed:
Add module for non-determinism to base library

This module allows us to write examples with non-determinism
in Haskell syntax.

just95 committed:
Add proof for tracing example #169

just95 committed:
Rewrite sharing tests in Haskell syntax	#169

just95 committed:
Remove `call` from test cases #169

just95 committed:
Fix code excerpts of pattern matching compiler #210

just95 committed:
Add examples for sharing and deep sharing

just95 committed:
Simplify QuickCheck implementation

just95 committed:
Add example that `id True` is `False` with sharing

just95 committed:
Specify for each handler how to interpret impure properties

just95 committed:
Add example for reflexivity of `(===)` and `(==)`

just95 committed:
Format code
